### PR TITLE
Test Case 13: Verify Product Quantity in Cart

### PR DESCRIPTION
**Pull request by HebaAbdelaziz911**

### CODE REVIEW

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            = By.id("quantity");
+
+    private final By addToCartbtn_Button
+            = By.xpath("//button [@class= \"btn btn-default cart\"]");
+
+    private final By viewCartbtn_button
+            = By.xpath("//a[@href=\"/view_cart\"]/u");
+
+    private final By productAdded
+            = By.xpath("//td[@class = \"cart_quantity\"]/button[@class=\"disabled\"]");
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java:28]: No need to add new line while initializing the By object here.
Also, I believe that we can enhance the xpaths here.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Step("Click on View Product")
+    public ProductQuantity clickOnViewProduct() {
+        driver.element().click(viewProductbtn_button);
+        new Actions(driver.getDriver()).doubleClick().perform();
+        return this;
+    }
+
+    @Step("Verify that Product details is opened")
+    public ProductQuantity verifyOnProductDetails(){
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java:43]: Validation methods should be in the last section at the bottom after the action methods.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        driver.element().assertThat(verifyDetailsExist);
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java:44]: Please don't forget to add .perform(); at the end of any assertion.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
+package com.gizasystems.automationexercise.pages;
+import com.shaft.driver.SHAFT;
+import io.qameta.allure.Step;
+import org.openqa.selenium.By;
+import org.openqa.selenium.interactions.Actions;
+
+public class ProductQuantity {
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java:7]: Please add "Page" at the end of the page class name.

**Review by MahmoudElSharkawyGS — CHANGES_REQUESTED** [review]
- Please note that Test data should be in the test data json file and should be passed from the test cases, not in the page class/methods.
- Please execute the pipeline and add the pipeline execution link here.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java (hunk 1 above)
Comment (HebaAbdelaziz911) [reply src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java]: Solved.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java (hunk 2 above)
Comment (HebaAbdelaziz911) [reply src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java]: Solved.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java (hunk 3 above)
Comment (HebaAbdelaziz911) [reply src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java]: Solved.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java (hunk 4 above)
Comment (HebaAbdelaziz911) [reply src/main/java/com/gizasystems/automationexercise/pages/ProductQuantity.java]: Solved.

**Comment by HebaAbdelaziz911** [comment]
https://github.com/MahmoudElSharkawyGS/AutomationExercisePractice/actions/runs/7949556018

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    // Constructor
+    public ProductQuantityPage(SHAFT.GUI.WebDriver driver){
+        this.driver = driver;
+    }
+
+    @Step("Click on View Product")
+    public ProductQuantityPage clickOnViewProduct() {
+        driver.element().click(viewProductbtn_button);
+        new Actions(driver.getDriver()).doubleClick().perform();
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java:33]: If this line is to handle the ads, please pull latest from master and try the `GoogleAlert.dismissAlert(driver, locator)` method.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    @Step("Add Product to Cart")
+    public ProductQuantityPage addProductToCart(){
+        driver.element().click(addToCartbtn_Button);
+        return this;
+    }
+
+    @Step("Verify Product Added to Cart with exact quantity")
+    public ProductQuantityPage verifyExactQuantityAddedToCart(String quantity){
+        driver.element().click(viewCartbtn_button);
+        driver.element().assertThat(productAdded).text().isEqualTo(quantity).perform();
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java:55]: We should not have action step and validation step at the same method. Please sperate them into 2 methods (1 to click on viewCartbtn_button and one to assert as is)

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
+package com.gizasystems.automationexercise.pages;
+import com.shaft.driver.SHAFT;
+import io.qameta.allure.Step;
+import org.openqa.selenium.By;
+import org.openqa.selenium.interactions.Actions;
+
+public class ProductQuantityPage {
+
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java:8]: Please remove this empty line here.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    private final By verifyDetailsExist = By.xpath("//span/Label");
+
+    private final By hoverBtn_button = By.id("quantity");
+
+    private final By addToCartbtn_Button = By.xpath("//button [@class= \"btn btn-default cart\"]");
+
+    private final By viewCartbtn_button = By.xpath("//a[@href=\"/view_cart\"]/u");
+
+    private final By productAdded = By.xpath("//td[@class = \"cart_quantity\"]/button[@class=\"disabled\"]");
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java:23]: - No need to have empty line between locators, please remove them.
- Please replace the " in your xpath to '

**Review by MahmoudElSharkawyGS — CHANGES_REQUESTED** [review]
Please check the comments and don't forget to pull latest from master branch and reexecute the pipeline.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java (hunk 5 above)
Comment (HebaAbdelaziz911) [reply src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java:33]: Solved.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java (hunk 6 above)
Comment (HebaAbdelaziz911) [reply src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java:55]: Solved.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java (hunk 7 above)
Comment (HebaAbdelaziz911) [reply src/main/java/com/gizasystems/automationexercise/pages/ProductQuantityPage.java:8]: Solved.

**Comment by HebaAbdelaziz911** [comment]
https://github.com/MahmoudElSharkawyGS/AutomationExercisePractice/actions/runs/7989786993
The new workflow link.

**Comment by HebaAbdelaziz911** [comment]
https://github.com/MahmoudElSharkawyGS/AutomationExercisePractice/actions/runs/8004153181
The new workflow

**Pull request merged**